### ISSUE TITLE
AI-powered Exam Preparation App: RecurLearn – The Recursive Exam Coach

### ISSUE THREAD

**Issue by lukaszkoziorowski** [lukaszkoziorowski/fibi-budget#1]
## Product Concept
Design and implement **RecurLearn**, an AI-powered exam-preparation application that uses the principle of recursive learning to continuously enhance both the learner's knowledge and the app's own teaching effectiveness.

---

## Objective
- Help users prepare for exams with adaptive learning.
- Recursively improve both user knowledge and the app's instructional strategies.

---

### Recursive Learning Loop
1. **Learn → Test → Analyze**
   - User studies a topic and takes a quiz.
   - The app records accuracy, time per question, confidence, and error types.
2. **Reflective Layer (Recursive Feedback)**
   - The app analyzes strengths, weaknesses, and confusion.
   - Updates the study plan, question difficulty, and teaching approach for the next cycle.
3. **Meta-Learning Layer**
   - Evaluates whether strategy changes improved performance.
   - Recursively refines algorithms and recommendations.
4. **Knowledge Graph Integration**
   - Dynamic knowledge graph links topics and adapts learning paths based on user struggles and conceptual relationships.

---

## System Architecture
- **User Learning Layer:** Lessons, quizzes, reflections.
- **Performance Analytics Engine:** Tracks scores, speed, and confidence.
- **Adaptive Learning Engine:** Adjusts difficulty, topics, and feedback style.
- **Meta-Learning Engine:** Refines its own effectiveness over time.
- **Knowledge Graph Engine:** Connects and updates related concepts.

---

## Bonus Features
- Reflective Journal: NLP-powered reasoning error detection.
- Recursive Gamification: Rewards for consistent improvement.
- Learning Style Adaptation: Detects visual/verbal/practical learning preferences.
- AI Tutor Mode: Experiments with explanation styles, learns what works best.

---

## Suggested Tech Stack
- **Frontend:** React Native or Flutter
- **Backend:** Node.js or Python (FastAPI)
- **ML/AI:** PyTorch or TensorFlow
- **Data Layer:** PostgreSQL, Neo4j for the knowledge graph
- **Reinforcement learning:** For adaptive feedback optimization

---

## UX/UI Inspiration
See references for UX design principles and quiz/journal flows:

<img alt="UX Design Foundations course dashboard" width="1280" src="https://github.com/user-attachments/assets/ab81cbf3-66a5-4fa7-a85a-be71f13906bc" />
<img alt="Quiz interface example" width="1280" src="https://github.com/user-attachments/assets/15a0fb31-3941-4c1e-82f5-c937e9d23bda" />
<img alt="Lesson and feedback screen" width="1280" src="https://github.com/user-attachments/assets/8d0814e0-0cba-4b00-8fd2-bbd746b61e6f" />

---

This issue is for the full concept, logic, and mechanism. Sub-issues should be created for engineering, knowledge graph, UI, and ML/AI development.
